src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -387,37 +387,34 @@ FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Start or
+                // Emit the general purpose scan loop. At this point, we always need TryMatchAtCurrentPosition.  If we have any
+                // information that will enable TryFindNextPossibleStartingPosition to help narrow down the search, we need it,
+                // too, but otherwise it can be skipped.
+                needsTryMatch = true;
+                needsTryFind =
+                    rm.Tree.FindOptimizations.FindMode != FindNextStartingPositionMode.NoSearch ||
+                    rm.Tree.FindOptimizations.MinRequiredLength != 0 ||
+                    rm.Tree.FindOptimizations.LeadingAnchor != RegexNodeKind.Unknown ||
+                    rm.Tree.FindOptimizations.TrailingAnchor != RegexNodeKind.Unknown;
+
+                writer.WriteLine("// Search until we can't find a valid starting position, we find a match, or we reach the end of the input.");
+                writer.Write("while (");
+                if (needsTryFind)
+                {
+                    writer.WriteLine("TryFindNextPossibleStartingPosition(inputSpan) &&");
+                    writer.Write("       ");
+                }
+                writer.WriteLine("!TryMatchAtCurrentPosition(inputSpan) &&");
                 writer.WriteLine($"       base.runtextpos != {(!rtl ? "inputSpan.Length" : "0")})");
                 using (EmitBlock(writer, null))
                 {
                     writer.WriteLine($"base.runtextpos{(!rtl ? "++" : "--")};");
-                }
-            }
-            else
-            {
-                // Emit the general purpose scan loop, with (possible) timeouts.
-                needsTryFind = needsTryMatch = true;
-                writer.WriteLine("// Search until we can't find a valid starting position, we find a match, or we reach the end of the input.");
-                using (EmitBlock(writer, "while (TryFindNextPossibleStartingPosition(inputSpan))"))
-                {
-                    // Check the timeout every time we run the whole match logic at a new starting location, as each such
-                    // operation could do work at least linear in the length of the input.
-                    EmitTimeoutCheckIfNeeded(writer, rm);
 
-                    using (EmitBlock(writer, $"if (TryMatchAtCurrentPosition(inputSpan) || base.runtextpos == {(!rtl ? "inputSpan.Length" : "0")})"))
-                    {
-                        writer.WriteLine("return;");
-                    }
-                    writer.WriteLine();
-
-                    writer.WriteLine($"base.runtextpos{(!rtl ? "++" : "--")};");
+                    // Check the timeout at least once per failed starting location, as finding the next location and
+                    // attempting a match at that location could do work at least linear in the length of the input.
+                    EmitTimeoutCheckIfNeeded(writer, rm, appendNewLineIfTimeoutEmitted: false);
```

Comment:
Can't this cause us to perform more than O(N) operations for matching? IIUC, previously we would try to find starting point (which could go the length of the input), check timeout, and then try match at current position (could go through the length of the input). This change seems to move the check timeout to after the trymatch, so I wonder if we should instead have timeout check both in between tryFind... and after trymatch?

Reply:
The find can do up to O(N).  The match can do up to O(N) without itself having a check.  2*O(N) is still O(N).

Reply:
I agree Big-O notation suggests this is still linear with the length of the input, but I thought the way we communicated timeout checks recently, we said we would do at most N operations between timeout checks, as opposed to using Big-O complexity for guaranteed timeout checks.

Reply:
fair enough.

Reply:
I do however see why we may want to do timeouts in look arounds and not in this particular case between tryFind and tryMatch. With lookarounds you could have a lot of nested lookarounds in the pattern causing you to walk the input multiple times, so timeout check is good to have there, while tryFind and tryMatch, would potentially only walk through the input twice.

Reply:
> That is also why I thought that lookarounds have an additional timeout check today, since one lookaround could potentially be walking through the input one more time, so in my mind this was kind of similar.

Each lookaround can result in doing O(N) work between timeout checks, so if you have M lookarounds, you could do O(M*N) work, and that includes a lookaround inside of a loop, so it doesn't actually require you to write M lookarounds.  For example, if you have `((?=a+)a)+` evaluated against a long string of 'a's, each iteration of the loop is going to examine the remainder of the input.

> The way I was thinking of defining "operations" in my mind was walking through the input counting as N operations, and after this change, we are possibly walking through the input twice, hence in my mind I saw this as doing more "operations" than N

This is why algorithmic complexity is helpful, as it cuts through the noise.  Each of those N operations could be a single bit check, it could be a single comparison, it could be a call to RegexRunner.CharInClass which could do a binary search on a large set of ranges, and so on... the number of instructions that compose it are widely variable.